plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php

```diff
@@ -273,27 +278,96 @@ protected function get_additional_fields_response( \WC_Order $order ) {
 	 * @return array
 	 */
 	protected function get_additional_fields_schema() {
-		$additional_fields_keys = $this->additional_fields_controller->get_additional_fields_keys();
-
-		$fields = $this->additional_fields_controller->get_additional_fields();
-
-		$additional_fields = array_filter(
-			$fields,
-			function( $key ) use ( $additional_fields_keys ) {
-				return in_array( $key, $additional_fields_keys, true );
-			},
-			ARRAY_FILTER_USE_KEY
-		);
+		$order_only_fields = $this->additional_fields_controller->get_order_only_fields();
 
 		$schema = [];
-		foreach ( $additional_fields as $key => $field ) {
-			$schema[ $key ] = [
+		foreach ( $order_only_fields as $key => $field ) {
+			$field_schema = [
 				'description' => $field['label'],
 				'type'        => 'string',
 				'context'     => [ 'view', 'edit' ],
-				'required'    => true,
+				'required'    => $field['required'],
 			];
+
+			if ( 'select' === $field['type'] ) {
+				$field_schema['enum'] = array_map(
+					function( $option ) {
+						return $option['value'];
+					},
+					$field['options']
+				);
+			}
+
+			if ( 'checkbox' === $field['type'] ) {
+				$field_schema['type'] = 'boolean';
+			}
+
+			$schema[ $key ] = $field_schema;
 		}
 		return $schema;
 	}
+
+	/**
+	 * Check if any additional field is required, so that the parent item is required as well.
+	 *
+	 * @return bool
+	 */
+	protected function is_additional_fields_required() {
+		$additional_fields_schema = $this->get_additional_fields_schema();
+		return array_reduce(
+			array_keys( $additional_fields_schema ),
+			function( $carry, $key ) use ( $additional_fields_schema ) {
+				return $carry || $additional_fields_schema[ $key ]['required'];
+			},
+			false
+		);
+	}
+
+	/**
+	 * Sanitize and format additional fields object.
+	 *
+	 * @param array            $fields Values being sanitized.
+	 * @param \WP_REST_Request $request The Request.
+	 * @return array
+	 */
+	public function sanitize_additional_fields( $fields, $request ) {
+		$fields = array_reduce(
+			array_keys( $fields ),
+			function( $carry, $key ) use ( $fields, $request ) {
+				$carry[ $key ] = rest_sanitize_value_from_schema( wp_unslash( $fields[ $key ] ), $request, $key );
```

Comment:
I think we may have to do it like this:

`$carry[ $key ] = rest_sanitize_value_from_schema( wp_unslash( $fields[ $key ] ), $this->get_properties()['additional_fields']['properties'][$key], $key );`

because inside `rest_sanitize_value_from_schema` the second argument (`$args`) expects to have a `type` to sanitize against.

We should also code this defensively and ensure the properties exist instead of directly accessing them as in my example.

However, this doesn't strip out script tags, should we be doing that here?

Reply:
This should be done.